File: bjxa.c
/*-
 * Copyright (C) 2018-2020  Dridi Boukelmoune
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "config.h"

#include <assert.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#include "bjxa.h"
#include "bjxa_priv.h"

static const char *progname;

static void
usage(FILE* file)
{

	fprintf(file,
	    "Usage: %s <action> [args...]\n"
	    "\n"
	    "Available actions:\n"
	    "\n"
	    "  help\n"
	    "    Show this message and exit.\n"
	    "\n"
	    "  decode [<xa file> [<wav file>]]\n"
	    "    Read an XA file and convert it into a WAV file.\n"
	    "\n"
	    "  encode [--bits <4|6|8>] [wav file> [<xa file>]]\n"
	    "    Read a WAV file and convert it into an XA file.\n"
	    "    The default number of bits per sample, when left\n"
	    "    unspecified is 6.\n"
	    "\n",
	    progname);
}

static noreturn int
cmd_fail(const char *msg)
{

	fprintf(stderr, "bjxa: %s\n", msg);
	usage(stderr);
	exit(EXIT_FAILURE);
}

static int
open_files(int argc, char * const *argv)
{

	if (argc == 0)
		return (0);

	if (strcmp("-", *argv) && freopen(*argv, "r", stdin) == NULL) {
		perror("Error");
		return (-1);
	}

	argc--;
	argv++;

	if (argc == 0)
		return (0);

	if (strcmp("-", *argv) && freopen(*argv, "w", stdout) == NULL) {
		perror("Error");
		return (-1);
	}

	return (0);
}

int
main(int argc, char * const *argv)
{

	int bits = -1;

	progname = *argv;
	argc--;
	argv++;

	if (argc == 0)
		cmd_fail("Missing an action");

	if (!strcmp("help", *argv)) {
		usage(stdout);
		return (EXIT_SUCCESS);
	}
	else if (!strcmp("decode", *argv)) {
		argc--;
		argv++;
		if (argc > 2)
			cmd_fail("Too many arguments");
		if (open_files(argc, argv) < 0 || decode(stdin, stdout) < 0)
			return (EXIT_FAILURE);
	}
	else if (!strcmp("encode", *argv)) {
		argc--;
		argv++;
		if (argc > 0 && !strcmp("--bits", *argv)) {
			argc--;
			argv++;
			if (argc == 0)
				cmd_fail("Missing number of bits per sample");
			if (strlen(*argv) == 1)
				bits = **argv - '0';
			if (bits != 4 && bits != 6 && bits != 8)
				cmd_fail("Invalid number of bits per sample");
			argc--;
			argv++;
		}
		else {
			bits = 6;
		}
		assert(bits == 4 || bits == 6 || bits == 8);
		if (argc > 2)
			cmd_fail("Too many arguments");
		if (open_files(argc, argv) < 0)
			return (EXIT_FAILURE);
		fprintf(stderr, "bjxa: Encoding not implemented\n");
		return (EXIT_FAILURE);
	}
	else {
		cmd_fail("Unknown action");
	}

	return (EXIT_SUCCESS);
}

File: dotnet/bjxa.cs
/*-
 * Copyright (C) 2018  Dridi Boukelmoune
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

using System;
using System.IO;

using static System.Diagnostics.Trace;

namespace bjxa {
	class CliArgumentException: ArgumentException {
		public CliArgumentException(string message): base(message) { }
	}

	class Cli {
		static Stream xa;
		static Stream wav;

		static void Usage(TextWriter sw) {
			sw.WriteLine(String.Join(Environment.NewLine,
			    "Usage: bjxa.exe <action> [args...]",
			    "",
			    "Available actions:",
			    "",
			    "  help",
			    "    Show this message and exit.",
			    "",
			    "  decode [<xa file> [<wav file>]]",
			    "    Read an XA file and convert it into a " +
			        "WAV file.",
			    ""));
		}

		static void OpenFiles(string[] args) {
			if (args.Length > 1 && args[1] != "-")
				xa = File.Open(args[1], FileMode.Open);
			else
				xa = Console.OpenStandardInput();

			if (args.Length > 2 && args[2] != "-")
				wav = File.Create(args[2]);
			else
				wav = Console.OpenStandardOutput();
		}

		static void Decode() {
			Decoder dec = new Decoder();
			Format fmt = dec.ReadHeader(xa);
			dec.WriteRiffHeader(wav);

			byte[] buf_xa = new byte[fmt.BlockSizeXa];
			short[] buf_pcm = new short[fmt.BlockSamples];
			long pcm_block = fmt.BlockSizePcm;

			while (fmt.Blocks > 0) {
				if (xa.Read(buf_xa, 0, buf_xa.Length) !=
				    buf_xa.Length)
					throw new IOException(
					    "Unexpected end of file.");

				int blocks = dec.Decode(buf_xa, buf_pcm);
				Assert(blocks == 1);

				pcm_block = Math.Min(pcm_block,
				    fmt.DataLengthPcm);

				fmt.WritePcm(wav, buf_pcm, pcm_block);
				fmt.DataLengthPcm -= pcm_block;
				fmt.Blocks--;
			}

			Assert(fmt.DataLengthPcm == 0);
		}

		static void Exec(string[] args) {
			if (args.Length == 0)
				throw new
				    CliArgumentException("Missing an action");

			if (args[0] == "help") {
				Usage(Console.Out);
				return;
			}

			if (args[0] == "decode") {
				if (args.Length > 3)
					throw new CliArgumentException(
					    "Too many arguments");
				OpenFiles(args);
				Decode();
				return;
			}

			throw new CliArgumentException("Unknown action");
		}

		static int Main(string[] args) {
			try {
				Exec(args);
				return (0);
			}
			catch (Exception e) {
				Console.Error.Write("bjxa: ");
				Console.Error.WriteLine(e.Message);
				if (e is CliArgumentException)
					Usage(Console.Error);
				return (1);
			}
			finally {
				if (xa != null)
					xa.Dispose();
				if (wav != null)
					wav.Dispose();
			}
		}
	}
}
